pkg/vault/contracts/TimelockAuthorizer.sol

```diff
@@ -0,0 +1,368 @@
+
+    constructor(address admin) {
+        GRANT_ACTION_ID = getActionId(TimelockAuthorizer.grantPermissions.selector);
+        REVOKE_ACTION_ID = getActionId(TimelockAuthorizer.revokePermissions.selector);
+        SCHEDULE_DELAY_ACTION_ID = getActionId(TimelockAuthorizer.scheduleDelayChange.selector);
+
+        _grantPermission(getActionId(TimelockAuthorizer.grantPermissions.selector), admin, EVERYWHERE);
+        _grantPermission(getActionId(TimelockAuthorizer.revokePermissions.selector), admin, EVERYWHERE);
+    }
+
+    /**
+     * @dev Tells the action ID for a certain function selector
+     */
+    function getActionId(bytes4 selector) public view override returns (bytes32) {
+        return keccak256(abi.encodePacked(bytes32(uint256(address(this))), selector));
+    }
+
+    /**
+     * @dev Tells the permission ID for action `actionId`, account `account` and target `where`
+     */
+    function permissionId(
+        bytes32 actionId,
+        address account,
+        address where
+    ) public pure returns (bytes32) {
+        return keccak256(abi.encodePacked(actionId, account, where));
+    }
+
+    /**
+     * @dev Tells whether `account` has explicit permission to perform action `actionId` in `where`
+     */
+    function hasPermission(
+        bytes32 actionId,
+        address account,
+        address where
+    ) public view returns (bool) {
+        return
+            isPermissionGranted[permissionId(actionId, account, where)] ||
+            isPermissionGranted[permissionId(actionId, account, EVERYWHERE)];
+    }
+
+    /**
+     * @dev Tells whether `account` can perform action `actionId` in `where`
+     */
+    function canPerform(
+        bytes32 actionId,
+        address account,
+        address where
+    ) public view override returns (bool) {
+        return (delaysPerActionId[actionId] > 0) ? account == address(this) : hasPermission(actionId, account, where);
+    }
+
+    /**
+     * @dev Sets a new delay for action `actionId`
+     */
+    function setDelay(bytes32 actionId, uint256 delay) external {
+        _require(msg.sender == address(this), Errors.SENDER_NOT_ALLOWED);
+        delaysPerActionId[actionId] = delay;
+        emit ActionDelaySet(actionId, delay);
+    }
+
+    /**
+     * @dev Schedules a delay change of `newDelay` for action `actionId`
+     */
+    function scheduleDelayChange(
+        bytes32 actionId,
+        uint256 newDelay,
+        address[] memory executors
+    ) external returns (uint256 scheduledActionId) {
+        require(newDelay <= MAX_DELAY, "DELAY_TOO_LARGE");
+        bytes32 scheduleDelayActionId = keccak256(abi.encodePacked(SCHEDULE_DELAY_ACTION_ID, actionId));
+        _require(hasPermission(scheduleDelayActionId, msg.sender, address(this)), Errors.SENDER_NOT_ALLOWED);
+
+        uint256 actionDelay = delaysPerActionId[actionId];
+        bytes memory data = abi.encodeWithSelector(this.setDelay.selector, actionId, newDelay);
```

Comment:
```suggestion
        bytes memory data = abi.encodeWithSelector(this.setDelay.selector, actionId, newDelay);
        // The delay change is scheduled to execute after the current delay for the action has elapsed. This is critical, as
        // otherwise it'd be possible to execute an action with a delay shorter than its current one by first changing it to a
        // smaller (or zero) value.
        uint256 actionDelay = delaysPerActionId[actionId];
```